great_expectations/data_context/store/expectations_store.py

```diff
@@ -190,20 +112,79 @@ def gx_cloud_response_json_to_object_dict(response_json: Dict) -> Dict:
         return suite_dict
 
     def _add(self, key, value, **kwargs):
+        value = self._add_ids_to_new_objects(value)
         try:
-            return super()._add(key=key, value=value, **kwargs)
+            result = super()._add(key=key, value=value, **kwargs)
+            if self.cloud_mode:
+                result = cast(GXCloudResourceRef, result)
+                value = self._add_cloud_ids_to_local_suite_and_expectations(
+                    local_suite=value,
+                    cloud_suite=result.response["data"]["attributes"]["suite"],
+                )
+            return result
         except gx_exceptions.StoreBackendError:
             raise gx_exceptions.ExpectationSuiteError(
                 f"An ExpectationSuite named {value.expectation_suite_name} already exists."
             )
 
     def _update(self, key, value, **kwargs):
+        value = self._add_ids_to_new_objects(value)
         try:
-            return super()._update(key=key, value=value, **kwargs)
-        except gx_exceptions.StoreBackendError:
+            result = super()._update(key=key, value=value, **kwargs)
+            if self.cloud_mode:
+                result = cast(GXCloudResourceRef, result)
+                value = self._add_cloud_ids_to_local_suite_and_expectations(
+                    local_suite=value,
+                    cloud_suite=result.response["data"]["attributes"]["suite"],
+                )
+        except gx_exceptions.StoreBackendError as exc:
+            # todo: this generic error clobbers more informative errors coming from the store
             raise gx_exceptions.ExpectationSuiteError(
                 f"Could not find an existing ExpectationSuite named {value.expectation_suite_name}."
+            ) from exc
+
+    def _add_ids_to_new_objects(self, suite: ExpectationSuite) -> ExpectationSuite:
+        if self.cloud_mode:
+            # cloud backend creates IDs, and we add them after persistence
+            return suite
+        # ensure suite has an ID
+        if not suite.get("ge_cloud_id"):
+            suite["ge_cloud_id"] = str(uuid.uuid4())
+        if isinstance(suite, ExpectationSuite):
+            key = "expectation_configurations"
+        else:
+            # this will be true if a serialized suite is provided
```

Comment:
the `migrator` code is passing a`dict` here. I think that's probably wrong, and would like to remove that code, but that feels out of scope for this work, and this fork is required to get tests to pass.

Reply:
I would like the signature to read suite: `dict | ExpectationSuite` before this merges

Reply:
unfortunately that change causes other typing issues. mypy seems convinced that the only thing ever passed to this object is a dict, even though the `CloudMigrator` is passing a dict. For now I'll leave this signature in place, but I'll add a ticket to remove the migrator as a follow up, and change this code to use dot notation. It would need to change at the latest when we migrate the suite object to pydantic, since bracket syntax will no longer work.